File: Assets/Resources/WeaponSystem/Scripts/WeaponSystem/Damage.cs
using UnityEngine;
using System.Collections;

public class Damage : DamageBase
{
    public bool Explosive;
    public float ExplosionRadius = 20;
    public float ExplosionForce = 1000;
	public bool HitedActive = true;
	public float TimeActive = 0;
	public bool RandomTimeActive;
	private float timetemp = 0;

	void Awake(){

	}

    private void Start()
    {
		timetemp = Time.time;
		if(RandomTimeActive)
			TimeActive = Random.Range(TimeActive/2f,TimeActive);

        if (!Owner || !Owner.GetComponent<Collider>()) return;
        Physics.IgnoreCollision(GetComponent<Collider>(), Owner.GetComponent<Collider>());
		
		
    }

    private void Update()
    {
		if(TimeActive>0){
			if(Time.time >= (timetemp + TimeActive)){
				Active();
			}
		}
    }

    public void Active()
    {
        if (Effect)
        {
            GameObject obj = (GameObject) Instantiate(Effect, transform.position, transform.rotation);
            Destroy(obj, LifeTimeEffect);
        }

        if (Explosive)
            ExplosionDamage();


        Destroy(gameObject);
    }

    private void ExplosionDamage()
    {
        Collider[] hitColliders = Physics.OverlapSphere(transform.position, ExplosionRadius);
        for (int i = 0; i < hitColliders.Length; i++)
        {
            Collider hit = hitColliders[i];
            if (!hit)
                continue;

            if (hit.gameObject.GetComponent<DamageManager>())
            {
                if (hit.gameObject.GetComponent<DamageManager>())
                {
                    hit.gameObject.GetComponent<DamageManager>().ApplyDamage(Damage,Owner);
                }
            }
            if (hit.GetComponent<Rigidbody>())
                hit.GetComponent<Rigidbody>().AddExplosionForce(ExplosionForce, transform.position, ExplosionRadius, 3.0f);
        }

    }

    private void NormalDamage(Collision collision)
    {
        if (collision.gameObject.GetComponent<DamageManager>())
        {
            collision.gameObject.GetComponent<DamageManager>().ApplyDamage(Damage,Owner);
        }
    }

    private void OnCollisionEnter(Collision collision)
    {
		if (collision.contacts [0].otherCollider.gameObject.CompareTag ("Player")) {
			collision.contacts [0].otherCollider.gameObject.SendMessage("OnDamage",8f,SendMessageOptions.DontRequireReceiver);
		}
		if(HitedActive){
        	if (collision.gameObject.tag != "Missile" && collision.gameObject.tag != this.gameObject.tag)
        	{
            	if (!Explosive)
                	NormalDamage(collision);
            	Active();
        	}
		}else{
			Destroy(this.gameObject);
		}
    }
}

File: Resources/Scripts/Bullet/Damage.js
var effect:GameObject;
var Owner:GameObject;
var Damage:int = 20;
var Explosive:boolean;
var ExplosionRadius:float = 20;
var ExplosionForce:float = 1000;
function Start () {
	var tf : Transform;
	tf = transform;
	while(tf.parent != null){
		tf = tf.parent;
		if(tf.GetComponent(BoxCollider) != null){
			Owner = tf.gameObject;		
		}
	}
	
	if(Owner){
		if(Owner.GetComponent.<Collider>()){
			Physics.IgnoreCollision(this.GetComponent.<Collider>(),Owner.GetComponent.<Collider>());
		}
	}
	
	transform.parent = null;
}
function Update(){
	
}
function Active(){
	if(effect){
   		var obj:GameObject = Instantiate(effect,this.transform.position,this.transform.rotation);
   		GameObject.Destroy(obj,3);
   	}
   	
   	if(Explosive)
   	ExplosionDamage();
   	  	
   	Destroy(this.gameObject,0.17f);
}

function ExplosionDamage(){


}


function NormalDamage(collision : Collision){
	/*
	if(collision.gameObject.GetComponent(DamageManager)){
		collision.gameObject.GetComponent(DamageManager).ApplyDamage(Damage);
	}*/	
}

function OnCollisionEnter(collision : Collision) {
	if(collision.gameObject.tag!="Particle" && collision.gameObject.tag!="Bullet"){
		if(!Explosive)
		NormalDamage(collision);
		Active();
	}
   	
}